mono/utils/mono-log-flight-recorder.c

```diff
@@ -0,0 +1,225 @@
+
+static void
+init (gpointer *out)
+{
+	MonoFlightRecorder *recorder = mono_flight_recorder_init (MAX_RECORDER_LOG_LEN, sizeof (LogMessage));
+	*out = (gpointer) recorder;
+}
+
+static void
+handle_command (gpointer state, gpointer payload, gboolean at_shutdown)
+{
+	MonoFlightRecorder *recorder = (MonoFlightRecorder *) state;
+	LogQueueEntry *entry = (LogQueueEntry *) payload;
+
+	switch (entry->command) {
+	case MONO_FLIGHT_RECORDER_APPEND:
+		mono_flight_recorder_append (recorder, &entry->message);
+
+#if 0
+		// Dump all messages on each append. This is an aggressive, slow
+		// debugging method. 
+
+		LogMessage messages [MAX_RECORDER_LOG_LEN];
+		LogQueueDumpRequest dump;
+		dump.messages = (LogMessage *) messages;
+		dump.num_messages = 0;
+		dump.max_num_messages = MAX_RECORDER_LOG_LEN;
+		mono_log_dump_recorder_internal (recorder, &dump);
+		fprintf (stderr, "%zu messages\n", dump.num_messages);
+
+		for (int i=0; i < dump.num_messages; i++)
+			fprintf (stderr, "\t(%d): %s\n", i, dump.messages [i].message);
+#endif
+
+		break;
+	case MONO_FLIGHT_RECORDER_DUMP:
+		fprintf (stderr, "Log received dump\n");
+		mono_log_dump_recorder_internal (recorder, entry->dump);
+		break;
+	default:
+		g_assert_not_reached ();
+	}
+}
+
+static void
+cleanup (gpointer state)
+{
+	mono_flight_recorder_free ((MonoFlightRecorder *) state);
+}
+
+static MonoUtilityThread *logger_thread;
+
+
+/**
+ * mono_log_open_recorder:
+ * \param path Unused
+ * \param userData Unused
+ * Open access to recorder 
+ */   
+void
+mono_log_open_recorder (const char *path, void *userData)
+{
+	MonoUtilityThreadCallbacks callbacks;
+	callbacks.early_init = NULL;
+	callbacks.init = init;
+	callbacks.command = handle_command;
+	callbacks.cleanup = cleanup;
+	logger_thread = mono_utility_thread_launch (sizeof (LogMessage), &callbacks, MONO_MEM_ACCOUNT_OTHER);
+}
+
+/**
+ * mono_log_write_recorder:
+ * \param domain Identifier string
+ * \param level Logging level flags
+ * \param format \c printf format string
+ * \param vargs Variable argument list
+ * Write data to recorder.
+ */
+void
+mono_log_write_recorder (const char *log_domain, GLogLevelFlags level, mono_bool hdr, const char *message)
+{
+	int small_id = mono_thread_info_get_small_id ();
+	if (small_id < 0) {
+#if MONO_PRINT_DROPPED_MESSAGES
+		fprintf (stderr, "Dropping message because thread not attached yet:\n\t%s\n", message);
+#endif
+		return;
+	} else if (level & G_LOG_LEVEL_ERROR) {
+		fprintf (stderr, "\nFatal Error Occured: %s\n\nHistory:\n", message);
+		mono_log_dump_recorder ();
+		abort();
+	} else if (!logger_thread->run_thread) {
+#if MONO_PRINT_DROPPED_MESSAGES
+		fprintf (stderr, "Dropping message because thread killed:\n\t%s\n", message);
+#endif
+		return;
+	}
+
+	LogQueueEntry entry;
+	entry.command = MONO_FLIGHT_RECORDER_APPEND;
+	g_snprintf ((gchar *) &entry.message.message, MAX_RECORDER_MSG_LEN, message);
+	mono_utility_thread_send (logger_thread, &entry);
```

Comment:
maybe I'm missing something obvious here, but why can't we append the message here into the ring buffer with the lockless stuff you do in the other thread? having a dedicated thread for this feels over-engineered.

Reply:
What lockless stuff? With the lockless queue? Making a lockless ring buffer with high scalability would be harder, I think. This is fairly fast and keeps things simple by reusing existing components. The ring buffer data structure itself is not lockless.

Reply:
I examined existing algorithms for lockless ring buffers, and found that the amount of memory allocated and freed was too much for us. I didn't want to add some big performance sink, just for a logger.

Reply:
If you consider the design of the profiler, and the necessity of similar threads with lockless queues in order to log into a data structure without slowing down mono, I think that this design looks a lot less radical. I went with a solution we've proven works elsewhere in the runtime, using the same APIs as that other implementation.